WebTransport should be able to send data before initial handshake roundtrip completes

achristensen07 opened the pull request:
#### 3e79d8b6c740989700d7a9b712995f7977813ba8
<pre>
WebTransport should be able to send data before initial handshake roundtrip completes
<a href="https://rdar.apple.com/158806467">rdar://158806467</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=298270">https://bugs.webkit.org/show_bug.cgi?id=298270</a>

Reviewed by Matthew Finkel.

Life is too short to wait for WebTransport.ready.  We&apos;ve got data to send.
Allowing this required changing the initialization to have a session immediately
then resolve the promise when it becomes ready.

This new setup method works pretty well most of the time, but to avoid calling
WebProcess::singleton().ensureNetworkProcessConnection() on a background thread,
we cache the connection in m_networkProcessConnection.  This introduces the possibility
that, if the network process crashes after we cache the connection, we would fail
to setup a new WebTransport in a worker.  I added a test that hits this case, and
I add a dispatchSync from the worker to the main thread to recover in this rare case.
In the common case, even setting up WebTransport from workers, we won&apos;t dispatchSync.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::initializeOverHTTP):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj.orig: Copied from Source/WebCore/WebCore.xcodeproj/project.pbxproj.
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/SocketProvider.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::initializeWebTransportSession):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::create):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::messageSenderDestinationID const):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
(WebKit::NetworkTransportSession::create): Deleted.
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::create):
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::createParameters):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::setupDatagramConnection):
(WebKit::NetworkTransportSession::setupConnectionHandler):
(WebKit::NetworkTransportSession::createStream):
(WebKit::NetworkTransportSession::receiveDatagramLoop):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebSocketProvider.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Source/WebKitLegacy/WebCoreSupport/LegacySocketProvider.cpp:
(LegacySocketProvider::initializeWebTransportSession):
* Source/WebKitLegacy/WebCoreSupport/LegacySocketProvider.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLED_CreateStreamsBeforeReady)):

Canonical link: <a href="https://commits.webkit.org/299715@main">https://commits.webkit.org/299715@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3628bab612df41ce4cd36f7249d0983328547726

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119964 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39657 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126290 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72037 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/12b89759-7953-4f48-ab44-61a41dabeb6b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121840 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40353 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48234 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/91097 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60410 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f3d3f039-7565-49f6-913e-5ad6b0dad05a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122916 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32226 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107563 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71653 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d7c496fb-6b64-4fb3-b553-26e4692ba771) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/119321 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31257 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25669 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69929 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101701 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129213 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46884 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35547 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/99723 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47250 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103752 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/99568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45013 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23028 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43505 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19072 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46746 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52452 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46212 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49561 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47898 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
```diff
@@ -423,7 +428,61 @@ WebTransportServer echoServer([](ConnectionGroup group) -> ConnectionTask {
 TEST(WebTransport, Worker)
+{
+    if (canLoadnw_webtransport_options_set_allow_joining_before_ready())
+        runWorkerTest(false);
```
Nit: seems like you could inline `runWorkerTest` here

sysrqb commented on Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
```diff
@@ -187,4 +187,5 @@ void WebTransportSession::destroyStream(WebCore::WebTransportStreamIdentifier id
 {
     send(Messages::NetworkTransportSession::DestroyStream(identifier, errorCode));
 }
+
```
nit

achristensen07 replied:
This one's intentional.  There was a missing newline.

sysrqb commented on Source/WebKit/NetworkProcess/cocoa/NetworkSoftLink.h:
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
nit: 2025

sysrqb commented on Source/WebCore/Modules/webtransport/WorkerWebTransportSession.cpp:
```diff
@@ -122,6 +122,7 @@ Ref<WritableStreamPromise> WorkerWebTransportSession::createOutgoingUnidirection
     ASSERT(!RunLoop::isMain());
     if (RefPtr session = m_session)
         return session->createOutgoingUnidirectionalStream();
+    ASSERT_NOT_REACHED();
```
Maybe add a log so we can debug these cases?

sysrqb commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -47,34 +48,32 @@ RefPtr<ThreadableWebSocketChannel> WebSocketProvider::createWebSocketChannel(Doc
+        if (!m_networkProcessConnection->isValid()) {
+            WorkQueue::mainSingleton().dispatchSync([protectedThis = Ref { *this }] {
+                ASSERT(RunLoop::isMain());
+                protectedThis->m_networkProcessConnection = WebProcess::singleton().ensureNetworkProcessConnection().connection();
```
Any concerns with raciness?

achristensen07 replied:
I should add m_networkProcessConnectionLock.

sysrqb commented on Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
```diff
@@ -40,30 +40,30 @@
+    auto identifier = WebTransportSessionIdentifier::generate();
+    return {
+        adoptRef(*new WebTransportSession(connection.copyRef(), WTFMove(client), identifier)),
+        connection->sendWithPromisedReply(Messages::NetworkConnectionToWebProcess::InitializeWebTransportSession(identifier, url, pageID, clientOrigin))->whenSettled(RunLoop::mainSingleton(), [] (auto&& result) {
```
If `connection` is invalid, this fail, right? Do we need a recovery path when the connection is reestablished?

achristensen07 replied:
We check that before getting here.  On the main thread we always ensure it is valid, on a worker thread we always check just before calling initialize, and we hop to the main thread an get a new one if it's not valid.

achristensen07 committed:
WebTransport should be able to send data before initial handshake roundtrip completes
rdar://158806467
https://bugs.webkit.org/show_bug.cgi?id=298270

Reviewed by Matthew Finkel.

Life is too short to wait for WebTransport.ready.  We've got data to send.
Allowing this required changing the initialization to have a session immediately
then resolve the promise when it becomes ready.

This new setup method works pretty well most of the time, but to avoid calling
WebProcess::singleton().ensureNetworkProcessConnection() on a background thread,
we cache the connection in m_networkProcessConnection.  This introduces the possibility
that, if the network process crashes after we cache the connection, we would fail
to setup a new WebTransport in a worker.  I added a test that hits this case, and
I add a dispatchSync from the worker to the main thread to recover in this rare case.
In the common case, even setting up WebTransport from workers, we won't dispatchSync.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::initializeOverHTTP):
* Source/WebCore/Modules/webtransport/WorkerWebTransportSession.h:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj.orig: Copied from Source/WebCore/WebCore.xcodeproj/project.pbxproj.
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/SocketProvider.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::initializeWebTransportSession):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::create):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::messageSenderDestinationID const):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
(WebKit::NetworkTransportSession::create): Deleted.
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::create):
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::createParameters):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::setupDatagramConnection):
(WebKit::NetworkTransportSession::setupConnectionHandler):
(WebKit::NetworkTransportSession::createStream):
(WebKit::NetworkTransportSession::receiveDatagramLoop):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebSocketProvider.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Source/WebKitLegacy/WebCoreSupport/LegacySocketProvider.cpp:
(LegacySocketProvider::initializeWebTransportSession):
* Source/WebKitLegacy/WebCoreSupport/LegacySocketProvider.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLED_CreateStreamsBeforeReady)):

Canonical link: https://commits.webkit.org/299715@main
